src/NumericalAlgorithms/Spectral/LogicalCoordinates.cpp

```diff
@@ -22,11 +23,63 @@ void logical_coordinates(
     const Mesh<VolumeDim>& mesh) {
   set_number_of_grid_points(logical_coords, mesh.number_of_grid_points());
   for (size_t d = 0; d < VolumeDim; ++d) {
-    const auto& collocation_points_in_this_dim =
-        Spectral::collocation_points(mesh.slice_through(d));
-    for (IndexIterator<VolumeDim> index(mesh.extents()); index; ++index) {
-      logical_coords->get(d)[index.collapsed_index()] =
-          collocation_points_in_this_dim[index()[d]];
+    switch (mesh.basis(d)) {
+      case Spectral::Basis::SphericalHarmonic: {
+        switch (mesh.quadrature(d)) {
+          case Spectral::Quadrature::Gauss: {
+            const size_t n_theta = mesh.extents(d);
+            std::vector<double> theta(n_theta + 1);
+            DataVector temp(2 * n_theta + 1);
+            auto work = gsl::make_span(temp.data(), n_theta);
+            auto unused_weights =
+                gsl::make_span(temp.data() + n_theta, n_theta + 1);
+
+            int err = 0;
+            gaqd_(static_cast<int>(n_theta), theta.data(),
+                  unused_weights.data(), work.data(),
+                  static_cast<int>(unused_weights.size()), &err);
+            if (UNLIKELY(err != 0)) {
+              ERROR("gaqd error " << err << " in LogicalCoordinates");
+            }
+            for (IndexIterator<VolumeDim> index(mesh.extents()); index;
+                 ++index) {
+              logical_coords->get(d)[index.collapsed_index()] =
+                  theta[index()[d]];
+            }
+            break;
+          }
+          case Spectral::Quadrature::Equiangular: {
+            const size_t n_phi = mesh.extents(d);
+            const double two_pi_over_n_phi = 2.0 * M_PI / n_phi;
+            for (IndexIterator<VolumeDim> index(mesh.extents()); index;
+                 ++index) {
+              logical_coords->get(d)[index.collapsed_index()] =
+                  two_pi_over_n_phi * index()[d];
+            }
+            break;
+          }
+          default:
+            ERROR(
+                "Quadrature must be Gauss or Equiangular for Basis "
+                "SphericalHarmonic");
+        }
+        break;
+      }
+      case Spectral::Basis::Chebyshev:
+        // fall through
+      case Spectral::Basis::Legendre:
+        // fall through
+      case Spectral::Basis::FiniteDifference: {
+        const auto& collocation_points_in_this_dim =
+            Spectral::collocation_points(mesh.slice_through(d));
```

Comment:
Move the computation of the collocation points one level deeper if possible, into `Spectral::collocation_points`. That way you also get the static caching for free.

Reply:
While it is possible, I don't think it makes sense to do so.  There are currently ten functions that use the macro `SPECTRAL_QUANTITY_FOR_MESH` and only `collocation_points` would make sense with Basis::SphericalHarmonic.  So since the branching is unique, why bury it and then require error handling for all the other functions?  Furthermore `collocation_points` calls `compute_collocation_points_and_weights` to get the collcation points, and I don't think there are any useful weights

Reply:
Ok, just two concerns: (1) it would be nice to get the static caching (not sure how important it actually is in practice for collocation points, certainly more important for diff matrices). (2) When calling `Spectral::collocation_points` it gives a "missing basis" error, but `logical_coordinates` with a 1D Ylm mesh will happily return the points. Moving the collocations points a level deeper would fix this inconsistency (I don't think it's a problem to move the missing-basis error from `get_spectral_quantity_for_mesh` a level deeper into the diff matrices etc).

Reply:
But then, as Larry said, all the other functions would still be inconsistent. I don't see any benefit to having the inconsistencies be in function defined in `Spectral.hpp`, I find it easier to have the difference in behavior at the `logical_coordinates` (which is not in `Spectral.hpp`). We should probably rename `Spectral.hpp` to `JacobiPolynomials.hpp` or `Basis1d.hpp` or something. `Spectral.hpp` is not a very clear name and is now clearly different from our 2d basis functions. I think the current implementation is the right avenue.
I'm not worried about caching of the logical coordinates at this point, I'm more concerned about getting the benefits of spherical shells into the code.

Reply:
Ok, I don't think it would be "inconsistent" to have the subset of functions in Spectral.hpp without Ylm implementation error out if a Ylm basis is requested. They already error when a Ylm basis is requested. Only `Spectral::collocation_points` is inconsistent because it errors with "missing basis" for a Ylm basis although `logical_coordinates` will return the collocation points just fine.

Anyway, the current implementation is fine with me. Maybe add a better error message to `Spectral::collocation_points` when a Ylm basis is requested or something like that.

Reply:
Ah, okay. I see. Yes, a better error message seems very reasonable. @kidder would you mind doing that?